src/lstm/recodebeam.cpp

```diff
@@ -88,15 +97,369 @@ void RecodeBeamSearch::Decode(const NetworkIO &output, double dict_ratio,
     }
   }
 }
+
+// determines top_n choices for given step and saves for later
+void RecodeBeamSearch::ComputeAndSaveTopN(const float *outputs, int num_outputs, int top_n) {
+
+  top_heap_.clear();
+  for (int i = 0; i < num_outputs; ++i) {
+    if (top_heap_.size() < top_n || outputs[i] > top_heap_.PeekTop().key()) {
+      TopPair entry(outputs[i], i);
+      top_heap_.Push(&entry);
+      if (top_heap_.size() > top_n) {
+        top_heap_.Pop(&entry);
+      }
+    }
+  }
+
+  std::vector<TopPair> topn_for_step;
+  topn_for_step.resize(top_heap_.size());
+  
+  int i = top_heap_.size();
+  while (!top_heap_.empty()) {
+    TopPair entry;
+    top_heap_.Pop(&entry);
+    TopPair save_entry(entry.key(), entry.data());
+    --i;
+    topn_for_step[i] = save_entry;  
+  }
+
+  save_topn_.push_back(topn_for_step);
+}
+
+// searches for potential diplopia cases and zeros scores
+// for all choices other than the highest one
+void RecodeBeamSearch::FindAndRemoveDiplopia(int width) {
+
+  int latest = -1;
+  int timestep = LocateDiplopia(width, latest);
+  while (timestep >= 0) {
+    latest = timestep;
+    RemoveChosenDiplopiaChar(width);
+    timestep = LocateDiplopia(width, latest);
+  }
+
+}
+
+// locates timestep (if any) at which apparent diplopia is occurring
+int RecodeBeamSearch::LocateDiplopia(int width, int latest) {
+
+  int retval = -1;
+
+  for (int t = latest + 1; t < width; ++t) {
+    std::vector<TopPair> topn_for_step = save_topn_[t];
+    int nbr_entries_topn = topn_for_step.size();
+    first_diplopia_char_ = -1;
+    second_diplopia_char_ = -1;
+    if (nbr_entries_topn >= 2) {
+      if (topn_for_step[0].data() != null_char_ && topn_for_step[1].data() != null_char_) {
+        first_diplopia_char_ = topn_for_step[0].data();
+        second_diplopia_char_ = topn_for_step[1].data();
+        DetermineDiplopiaCharToRemove(width, t);
+        if (diplopia_gap_ <= kMaxDiplopiaGap) {
+          retval = t;
+        }
+      }
+    }
+    if (retval >= 0) {
+      break;
+    }
+  }
+
+  return retval;
+}
+
+void RecodeBeamSearch::DetermineDiplopiaCharToRemove(int width, int timestep) {
+
+  DetermineDiplopiaDimensions(width, timestep, first_diplopia_char_);
+  int start_first_diplopia = start_diplopia_; 
+  int end_first_diplopia = end_diplopia_;
+  float first_diplopia_max = diplopia_max_;
+  int first_diplopia_max_timestep = diplopia_max_timestep_;
+  
+  DetermineDiplopiaDimensions(width, timestep, second_diplopia_char_);
+  int start_second_diplopia = start_diplopia_;
+  int end_second_diplopia = end_diplopia_;
+  float second_diplopia_max = diplopia_max_;
+  int second_diplopia_max_timestep = diplopia_max_timestep_;
+
+  if (first_diplopia_max >= second_diplopia_max) {
+    chosen_diplopia_char_ = second_diplopia_char_;
+    start_diplopia_ = start_second_diplopia;
+    end_diplopia_ = end_second_diplopia;
+    diplopia_max_ = second_diplopia_max;
+  }
+  else {
```

Comment:
The tesseract coding style is different. Please try to use the same style as in the rest of the tesseract code.

Reply:
I have updated the code for the if/else style.   If there are other stylistic changes needed, please let me know what they are and I will attempt to address them.

As far as the int vs auto, I have not made any changes for that as yet.  I do see in the original code in areas that I have not modified there are many examples where auto is not used.

Thanks,

Dave